Add server side TLS certificate expiry tracking and INFO telemetry

YiwenZhang12 opened the pull request:
regarding issue: https://github.com/valkey-io/valkey/issues/2784


- Parse TLS certificate expiry/serials and refresh cached values at startup and on TLS reconfigure; 

- expose INFO # TLS fields so monitoring can alert on impending expiry.


INFO:
```
# TLS
tls_server_cert_serial:826F01147AC63276
tls_server_cert_expires_in_seconds:123456
tls_client_cert_serial:826F01147AC63276
tls_client_cert_expires_in_seconds:123456
tls_ca_cert_serial:11A566...
tls_ca_cert_expires_in_seconds:987654
```

Commit:
issue/2784

Signed-off-by: Yiwen Zhang <yiwen_zhang@apple.com>

YiwenZhang12 commented:
Hi all, this PR is still missing tests. Before I add the test coverage, I wanted to check whether the approach and implementation look correct to you. If the direction is acceptable, I will proceed with writing the tests. Thanks!

Commit:
update comments

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

madolson commented on src/server.c:
```diff
@@ -5919,6 +5951,17 @@ sds genValkeyInfoString(dict *section_dict, int all_sections, int everything) {
+        if (sections++) info = sdscat(info, "\r\n");
+        info = sdscatprintf(info,
+                            "# TLS\r\n"
+                            "tls_enabled:%s\r\n"
```
I don't think we need to include basic information here like port and enabled, those can be easily extracted from configs, and it's better to keep this section paired down. I would also include the serial here though. Some folks like to verify the correct serial is stored.

soloestoy commented:
good feature

zuiderkwast commented on src/server.c:
```diff
@@ -5919,6 +5951,17 @@ sds genValkeyInfoString(dict *section_dict, int all_sections, int everything) {
+                            "tls_enabled:%s\r\n"
+                            "tls_server_cert_expires_in_seconds:%lld\r\n",
+                            (server.tls_port || server.tls_replication || server.tls_cluster) ? "yes" : "no",
+                            server.tls_server_cert_expires_in_seconds);
```
The variable `server.tls_server_cert_expires_in_seconds ` is updated by the cron job only once every hour. If the INFO command is called multiple times within one hour, it will show the same number of remaining seconds, even if some seconds or minutes have passed between the INFO calls. This is confusing and incorrect.

Let's instead store the expiration time of the certificate in the server variable and calculate the remaining seconds when we print the INFO field.

```suggestion
                            server.tls_server_cert_expire_time - server.unixtime);
```

We can read the expire time from the certificate when it is configured. We don't need to do it every hour.

zuiderkwast commented on src/config.c:
```diff
@@ -2745,6 +2745,7 @@ static int applyTlsCfg(const char **err) {
         *err = "Unable to update TLS configuration. Check server logs.";
         return 0;
     }
+    tlsUpdateServerCertInfo();
```
We can fetch the cert expire time inside tlsConfigure in tls.c instead. (It is called by connTypeConfigure().)

The tls.c file already accesses the global server object so we can do it for this information too.

zuiderkwast commented on src/server.c:
```diff
@@ -3110,6 +3117,31 @@ void initListeners(void) {
 }
 
+/* Refresh cached TLS certificate expiration metadata and emit logs */
+void tlsUpdateServerCertInfo(void) {
```
Let's move the code to update cert expiry time to tls.c.

In the cron job, we only need to check the time and log the warning every hour.

YiwenZhang12 committed:
Track TLS cert expiry timestamp

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 commented:
@zuiderkwast  Thanks for the detailed feedback! I've made the changes you suggested:

- Expiry tracking moved into tls.c
- INFO now reports the actual remaining seconds
- Cron job just emits the warning
- Added tests to cover the new behavior

Let me know if this looks better!

madolson commented on src/server.c:
```diff
@@ -5925,6 +5931,22 @@ sds genValkeyInfoString(dict *section_dict, int all_sections, int everything) {
+        info = sdscatprintf(info,
+                            "# TLS\r\n"
+                            "tls_enabled:%s\r\n"
+                            "tls_server_cert_expires_in_seconds:%lld\r\n",
```
Let's add the client and root CAs as well.

madolson commented on src/tls.c:
```diff
@@ -193,6 +195,162 @@ static void tlsCleanup(void) {
+        long long days_remaining = (seconds_remaining + 86400 - 1) / 86400;
+        if (days_remaining < 1) days_remaining = 1;
+        serverLog(LL_NOTICE, "TLS server certificate expiring in %lld days", days_remaining);
+    }
```
Is this needed? The warning seems useful, but this one doesn't seem useful.

madolson reviewed (commented):
Broadly looks good. We don't have a majority of TSC in the meeting, but madolson, zuiderkwast, and Ran conditionally approved it. If no one else comments in two weeks we can mark it major decision approved.

YiwenZhang12 commented:
Hi @madolson Thanks for the review and the helpful comments!

I removed the redundant `tls_enabled` flag from `INFO tls`, so the section now focuses only on certificate metadata. For each loaded TLS cert (server, client, and CA), it includes both serial number and an “expires in seconds” value.

I also added a configurable tls-cert-warning-days (default 7) to control when daily expiry log start. If a cert is already expired, we still log a warning regardless of the threshold. All three certs are included.

example:
INFO
```
# TLS
tls_server_cert_serial:826F01147AC63276
tls_server_cert_expires_in_seconds:123456
tls_client_cert_serial:826F01147AC63276
tls_client_cert_expires_in_seconds:123456
tls_ca_cert_serial:11A566...
tls_ca_cert_expires_in_seconds:987654
```

LOG:
```
[NOTICE] TLS server certificate (serial 826F01147AC63276) expiring in 5 days
[WARNING] TLS CA certificate (serial 11A566...) has EXPIRED
```

YiwenZhang12 committed:
tls: track CA bundle expiry

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
tests: generate multi-CA bundle

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
tests: cleanup

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
Add TLS CA-dir KPI reporting

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
Tidy cert generation helpers

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
Merge branch 'unstable' into feature/server_cert

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
tests: append TLS expiry counter tests

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
refactor

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

YiwenZhang12 committed:
test: check tls cert serial on reload

Signed-off-by: Yiwen Zhang <zhangyiwen1221@gmail.com>

Commit:
Resolve merge conflict

Signed-off-by: Yiwen Zhang <yiwen_zhang@apple.com>

Commit:
Update TLS tests and cert generation

Signed-off-by: Yiwen Zhang <yiwen_zhang@apple.com>

madolson commented on tests/unit/tls.tcl:
```diff
@@ -437,3 +448,257 @@ start_server {tags {"tls"}} {
+                set info [r info tls]
+                if {![regexp {tls_client_cert_serial:([^\r\n]+)} $info -> client_serial]} {
+                    fail "INFO tls missing tls_client_cert_serial"
+                }
```
```suggestion
                set info [r info tls]
                assert_not_equal [getInfoProperty $info tls_client_cert_serial] {}
```
Maybe? There are a bunch of existing helpers to do info parsing.

madolson commented on tests/unit/tls.tcl:
```diff
@@ -437,3 +448,257 @@ start_server {tags {"tls"}} {
+        test {TLS: INFO tls shows none for missing client cert} {
+            start_server [list overrides [list tls-auth-clients no] \
+                               omit [list tls-client-cert-file tls-client-key-file]] {
+                set s [valkey [srv 0 host] [srv 0 port] 0 1]
```
Does this client do anything?

YiwenZhang12 replied:
good catch. removed unused client

madolson commented on src/tls.c:
```diff
@@ -190,13 +197,354 @@ static void tlsCleanup(void) {
+/* Convert ASN1_TIME into a UTC tm plus a timezone offset (seconds). */
+static int tlsAsn1TimeToTm(const ASN1_TIME *time, struct tm *tm, int *tz_off) {
+    if (!time || !tm) return 0;
+#if OPENSSL_VERSION_NUMBER >= 0x10002000L || defined(LIBRESSL_VERSION_NUMBER)
```
OpenSSL 1.0.2 was deprecated in 2020. I think we can just assume this function exists. We can also just return 0 for older versions.

A lot of the code seems to be trying to support the fallback case, and I don't see a super compelling reason to support super old deprecated versions.

YiwenZhang12 replied:
sure. Dropped old fallback parsing; rely on ASN1_TIME_to_tm and return 0 for older versions.

madolson commented on tests/unit/tls.tcl:
```diff
@@ -437,3 +448,257 @@ start_server {tags {"tls"}} {
+                assert_morethan $exp_value 0
+            }
+
+            after 1200
```
Is this number based off of anything? Can we do some wait for a value instead?

YiwenZhang12 replied:
Replaced with `wait_for_condition`

madolson commented on src/config.c:
```diff
@@ -2768,7 +2769,7 @@ static int applyTLSPort(const char **err) {
         return 0;
     }
-
+    tlsResetCertInfo();
```
Why are we clearing the certs here? An end user might just be changing the TLS port, which would clear all the values?

YiwenZhang12 replied:
Thanks for the question @madolson . The intent is for the `INFO tls_*` fields to reflect active TLS state, not configured files. When TLS is disabled we clear them to avoid showing stale values. `tlsResetCertInfo()` is a no‑op when TLS is enabled (it early‑returns)
`if (server.tls_port || server.tls_replication || server.tls_cluster) return;`
, so changing the TLS port won’t clear anything.

madolson commented on src/server.c:
```diff
@@ -5984,6 +5997,42 @@ sds genValkeyInfoString(dict *section_dict, int all_sections, int everything) {
+                            "tls_client_cert_expires_in_seconds:%lld\r\n"
+                            "tls_ca_cert_serial:%s\r\n"
+                            "tls_ca_cert_expires_in_seconds:%lld\r\n"
+                            "tls_ca_cert_count:%d\r\n",
```
What is the use case for `tls_ca_cert_count`?

YiwenZhang12 replied:
good point. removed the field keep KPI focused and reduce noise

madolson commented on src/tls.c:
```diff
@@ -190,13 +197,354 @@ static void tlsCleanup(void) {
     }
+    tlsClearCertInfo(&server.tls_server_cert_expire_time, &server.tls_server_cert_serial);
+    tlsClearCertInfo(&server.tls_client_cert_expire_time, &server.tls_client_cert_serial);
+    tlsClearCACertInfo();
```
```suggestion
    tlsResetCertInfo();
```

YiwenZhang12 replied:
those three calls are now consolidated into tlsClearAllCertInfo() and both TLS and non‑TLS paths call that helper.

madolson commented on src/tls.c:
```diff
@@ -190,13 +197,354 @@ static void tlsCleanup(void) {
+    tlsClearCACertInfo();
+}
+
+/* Cache the certificate expiration timestamp for later INFO reporting. */
```
This function itself doesn't do any caching, it just extracts the serial out.

madolson commented on src/tls.c:
```diff
@@ -1620,6 +1977,30 @@ int RedisRegisterConnectionTypeTLS(void) {
 
 #else /* USE_OPENSSL */
 
+void tlsResetCertInfo(void) {
```
Isn't this almost the exact same as the TLS version?

YiwenZhang12 replied:
Good point. We already pulled the shared clearing logic into tlsClearAllCertInfo(), so the duplicated code is minimal now. The remaining tlsResetCertInfo() implementations live in different compile paths and have slightly different semantics (TLS build has the guard; non‑TLS always clears).

madolson reviewed (commented):
So, only high level thing is I'm still not very convinced the log for the warning is a great config, since I think people won't really watch for those and I'm also not clear about needing a metric for number of parsed certificates. Other than that the code mostly looks good, mostly just suggestions for simplication. Looks close to be being merged.

Commit:
remove log & code refactor

Signed-off-by: Yiwen Zhang <yiwen_zhang@apple.com>

Commit:
refacor and code cleanup

Signed-off-by: Yiwen Zhang <yiwen_zhang@apple.com>

enjoy-binbin commented on src/server.c:
```diff
@@ -5984,6 +5991,40 @@ sds genValkeyInfoString(dict *section_dict, int all_sections, int everything) {
+            tls_ca_seconds_remaining = server.tls_ca_cert_expire_time - (long long)server.unixtime;
+            if (tls_ca_seconds_remaining < 0) tls_ca_seconds_remaining = 0;
+        }
+        info = sdscatprintf(info,
```
we can use the new FMTARGS here, like
```
        info = sdscatprintf(
            info,
            "# Clients\r\n" FMTARGS(
```

YiwenZhang12 replied:
Thanks for taking a look @enjoy-binbin.  I’ve updated the INFO TLS block to use FMTARGS. Could you please take another look when you have a moment?

Commit:
using FMTARGS

Signed-off-by: Yiwen Zhang <yiwen_zhang@apple.com>

madolson reviewed (approved):
I'm ok with this. We talked during todays meeting about adding a log when the cert is actually expired, but I'm fine with or without that.
